src/main/java/uk/nhs/tis/trainee/actions/service/ActionService.java

```diff
@@ -274,6 +277,49 @@ public Optional<ActionDto> updateAction(CojReceivedEvent event) {
     }
   }
 
+  /**
+   * Updates the action associated with the given Form update event.
+   *
+   * @param event The form update event containing the data to update the action.
+   * @return An Optional containing the updated ActionDto, or empty if no action was updated.
+   */
+  public Optional<ActionDto> updateAction(FormUpdateEvent event) {
+    log.info("Updating action for form updated event: {}", event);
+    ActionType formAction = ActionType.getFormActionType(event.formType());
+    if (event.traineeId() == null || formAction == null || event.formContentDto() == null
+        || event.formContentDto().get(FORM_PROGRAMME_MEMBERSHIP_ID_FIELD) == null) {
+      log.warn("No Form data provided in the event.");
+      return Optional.empty();
+    }
+
+    String pmUuid = event.formContentDto().get(FORM_PROGRAMME_MEMBERSHIP_ID_FIELD).toString();
+    List<Action> existingActions = repository.findByTraineeIdAndTisReferenceInfo(
+        event.traineeId(), pmUuid, PROGRAMME_MEMBERSHIP.toString());
+    Optional<Action> existingAction = existingActions.stream()
+        .filter(a -> a.type().equals(formAction)).findFirst();
```

Comment:
The problem here is if the trainee has submitted more than one FormR PartA (for example) for a given PM. If one is unsubmitted the status of the action for signing-formr-part-a will be indeterminate: it could stay complete or become incomplete.